File: Ch5/01.write/write0.c
/*
 * write0.c
 * purpose: send messages to another terminal
 * method: open the other terminal for output then
 *         copy from stdin to that terminal
 * shows: a terminal is just a file supporting regular i/o
 * usage: write0 ttyname
*/ 

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <fcntl.h>
#include <string.h>

int main(int ac, char *av[]) {
    int fd = 0;
    char buf[BUFSIZ];
    
    // check args
    if (ac != 2) {
        fprintf(stderr, "Usage: write0 ttyname\n");
        exit(-1);
    }

    // open devices
    fd = open(av[1], O_WRONLY);
    if (fd == -1) {
        perror(av[1]);
        exit(-1);
    }

    // loop until EOF on input
    while (fgets(buf, BUFSIZ, stdin) != NULL) {
		printf("one iteration executed\n");
        if (write(fd, buf, strlen(buf)) == -1) {
            break;
        }
		printf("write one piece of contents\n");
    }
    close(fd);
}

File: Ch5/01.write/write0.rs
use std::env::args;
use std::process;
use std::io::{stdin, Stdin, Write, BufRead, BufReader};
use std::fs::File;


fn main() {
    let av: Vec<String> = args().collect();

    // check args
    if av.len() != 2 {
        eprintln!("Usage: write0 ttyname");
        process::exit(-1);
    }
   
    let mut tty: File = match File::create(av[1].as_str()) {
        Ok(f) => f,
        Err(e) => {
            eprintln!("Cannot open {}: {}", av[1].as_str(), e);
            process::exit(-1);
        }
    };
    
    let mut buf: Vec<u8> = Vec::new();
    let input: Stdin = stdin();
    let mut buf_input: BufReader<Stdin> = BufReader::new(input);

    while let Ok(n) = buf_input.read_until(10, &mut buf) {
        if n==0 {
            break;
        }

        println!("one iteration executed");
        if tty.write(buf.as_slice()).is_err() {
            break;
        }
        println!("write one piece of contents");
        buf.clear();
    }
}
